doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+a part of 2FA. Although 2FA reduces the risk of account hijacking, the benefits
+of application-specific passwords still improve security in this case as well.~~
+
+### Use as a fallback mechanism when 2FA is activated, but the client
+    application does not support it
+
+This is almost certainly not needed, as authentication is managed by FreeIPA,
+which obviously supports 2FA, and client applications only consume the result of
+the authentication. And again, 2FA is out of scope of this extension.
+
+### Use as part of RADIUS and Hardened Password authentication mechanisms
+
+These may work, as the code handling these authentication mechanisms is
+independent from the code handling the password(s), but it will not be a goal
+for this extension nor will it be tested.
+
+~~As both of these authentication mechanisms include sending a password for
+authentication, they could also be used with application-specific passwords.
+As pure RADIUS provides only weak protection of the sent password, it might
+be even more desirable than in the case of traditional password
+authentication.~~
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+Application-specific passwords will be stored in a new LDAP subtree in the form
+```
+object,cn=appname,cn=apps,$SUFFIX.
+```
+Every application will have its own subtree in which users will be able to
+create their objects if they will want to use that application - this means
+objects containing application-specific passwords will not be grouped by users,
+but by applications. There will be one object for each application-specific
+password of each user; one user will be able to have one password for a given
+application. The object containing an application-specific password will have
+the following attributes:
+* `krbprincipalkey` - will hold the application-specific password itself
+* `krbprincipalname` - two-component Kerberos principal (see below)
+* backlink to the owning user (that created it) - this backlink must be
+  maintained; inspiration can be seen in working with hostgroups
+
+The definition of the new object and its attributes holding application-specific
+passwords and related information, as well as operations with them (add, modify,
+delete) will be contained in a new file - `ipaserver/plugins/appspecificpw.py`.
+
+Names of the newly created objects in the new subtree will somehow have to be
+validated and enforced (possibly, but not necessarily using an LDAP plugin), so
+that the user will not be able to create a principal looking like a host or a
+service, or even set a password for another user.
+
+The existing mechanism for generating passwords will be made available for use
+for application-specific passwords as well. Of course, the user will be able to
+set application-specific passwords as usual (having to fulfil password policies
+in place).
+
+Password reset functionality must be extended to facilitate multiple passwords
```

Comment:
Do we actually need the feature? Google's app password feature is much simpler. It only lets you generate an app-specific password or delete an app-specific password. There is no way to set, reset, or retrieve an app password.

I suggest to limit the feature to generate, list, and delete an app password for now.

Reply:
So I really think we should have only one password policy, and that is length based, 20 chars or more. Nothing else.
As for reset, we need no such thing, a user should be able to simply delete an app password object and recreate a new one.
Only the user proper password should be considered for password resets.

Reply:
All right, there will be no possibility to reset an app-specific password. What is therefore left TBD, is how app-specific passwords will be created and managed, and also depending on that, what password policies will apply to them, if any. I will ask all outstanding questions at the end of the discussion.